File: exp-platform/windows/dotnet/GUIForCLIWindows.Core/BundleIconMap.cs
using System.Globalization;

namespace GUIForCLIWindows.Core;

public sealed class BundleIconMap
{
    public const string SfSymbolsSource = "sf-symbols";
    public const string WindowsSource = "windows";
    public const string BootstrapSource = "bootstrap";
    public const string EmojiSource = "emoji";

    public static BundleIconMap Empty { get; } = new(
        new Dictionary<string, IReadOnlyDictionary<string, string>>(StringComparer.Ordinal));

    public BundleIconMap(IReadOnlyDictionary<string, IReadOnlyDictionary<string, string>> sources)
    {
        Sources = sources;
    }

    public IReadOnlyDictionary<string, IReadOnlyDictionary<string, string>> Sources { get; }

    public string? Resolve(string source, string? key)
    {
        if (string.IsNullOrWhiteSpace(key))
        {
            return null;
        }

        return Sources.TryGetValue(source, out var values)
            && values.TryGetValue(key.Trim(), out var value)
            ? value
            : null;
    }

    public BundleIconMap Merge(BundleIconMap overrides)
    {
        var merged = Sources.ToDictionary(
            pair => pair.Key,
            pair => new Dictionary<string, string>(pair.Value, StringComparer.Ordinal),
            StringComparer.Ordinal);
        foreach (var (source, values) in overrides.Sources)
        {
            if (!merged.TryGetValue(source, out var target))
            {
                target = new Dictionary<string, string>(StringComparer.Ordinal);
                merged[source] = target;
            }

            foreach (var (key, value) in values)
            {
                target[key] = value;
            }
        }

        return new BundleIconMap(merged.ToDictionary(
            pair => pair.Key,
            pair => (IReadOnlyDictionary<string, string>)pair.Value,
            StringComparer.Ordinal));
    }

    public static BundleIconMap Parse(string text)
    {
        var sources = new Dictionary<string, Dictionary<string, string>>(StringComparer.Ordinal);
        string? currentSource = null;
        var lines = text.Replace("\r\n", "\n", StringComparison.Ordinal).Split('\n');
        for (var index = 0; index < lines.Length; index++)
        {
            var rawLine = lines[index];
            var lineNumber = index + 1;
            var line = rawLine.Trim();
            if (line.Length == 0 || line.StartsWith('#'))
            {
                continue;
            }

            if (line.StartsWith('[') && line.EndsWith(']'))
            {
                currentSource = line[1..^1].Trim();
                if (currentSource.Length == 0)
                {
                    throw InvalidLine(lineNumber, rawLine);
                }

                sources.TryAdd(currentSource, new Dictionary<string, string>(StringComparer.Ordinal));
                continue;
            }

            if (currentSource is null)
            {
                throw InvalidLine(lineNumber, rawLine);
            }

            var equals = FindUnescapedEquals(line);
            if (equals < 0)
            {
                throw InvalidLine(lineNumber, rawLine);
            }

            var key = UnquoteKey(line[..equals].Trim());
            var rawValue = line[(equals + 1)..].TrimStart();
            sources[currentSource][key] = ParseStringValue(rawValue, lineNumber, rawLine);
        }

        return new BundleIconMap(sources.ToDictionary(
            pair => pair.Key,
            pair => (IReadOnlyDictionary<string, string>)pair.Value,
            StringComparer.Ordinal));
    }

    private static int FindUnescapedEquals(string line)
    {
        var escaped = false;
        var inString = false;
        for (var index = 0; index < line.Length; index++)
        {
            var character = line[index];
            if (escaped)
            {
                escaped = false;
                continue;
            }

            if (character == '\\')
            {
                escaped = true;
                continue;
            }

            if (character == '"')
            {
                inString = !inString;
                continue;
            }

            if (!inString && character == '=')
            {
                return index;
            }
        }

        return -1;
    }

    private static string UnquoteKey(string key) =>
        key.StartsWith('"') && key.EndsWith('"') ? key[1..^1] : key;

    private static string ParseStringValue(string rawValue, int lineNumber, string rawLine)
    {
        if (!rawValue.StartsWith('"'))
        {
            throw InvalidLine(lineNumber, rawLine);
        }

        var escaped = false;
        for (var index = 1; index < rawValue.Length; index++)
        {
            var character = rawValue[index];
            if (escaped)
            {
                escaped = false;
                continue;
            }

            if (character == '\\')
            {
                escaped = true;
                continue;
            }

            if (character == '"')
            {
                var trailing = rawValue[(index + 1)..].Trim();
                if (trailing.Length > 0 && !trailing.StartsWith('#'))
                {
                    throw InvalidLine(lineNumber, rawLine);
                }

                return Unescape(rawValue[1..index], lineNumber, rawLine);
            }
        }

        throw InvalidLine(lineNumber, rawLine);
    }

    private static string Unescape(string value, int lineNumber, string rawLine)
    {
        var result = new List<char>();
        for (var index = 0; index < value.Length; index++)
        {
            var character = value[index];
            if (character != '\\')
            {
                result.Add(character);
                continue;
            }

            index += 1;
            if (index >= value.Length)
            {
                throw InvalidLine(lineNumber, rawLine);
            }

            var escaped = value[index];
            switch (escaped)
            {
                case 'n':
                    result.Add('\n');
                    break;
                case 'r':
                    result.Add('\r');
                    break;
                case 't':
                    result.Add('\t');
                    break;
                case '"':
                    result.Add('"');
                    break;
                case '\\':
                    result.Add('\\');
                    break;
                case 'u':
                case 'U':
                    var length = escaped == 'u' ? 4 : 8;
                    if (index + length >= value.Length)
                    {
                        throw InvalidLine(lineNumber, rawLine);
                    }

                    var hex = value.Substring(index + 1, length);
                    if (!int.TryParse(hex, NumberStyles.HexNumber, CultureInfo.InvariantCulture, out var codePoint))
                    {
                        throw InvalidLine(lineNumber, rawLine);
                    }

                    result.AddRange(char.ConvertFromUtf32(codePoint));
                    index += length;
                    break;
                default:
                    throw InvalidLine(lineNumber, rawLine);
            }
        }

        return new string([.. result]);
    }

    private static InvalidOperationException InvalidLine(int lineNumber, string rawLine) =>
        new($"Invalid icon map TOML at line {lineNumber}: {rawLine}");
}

File: platform/apple/shared/Sources/GUIForCLICore/Icons/BundleIconMap.swift
import Foundation

public struct BundleIconMap: Equatable, Sendable {
  public static let sfSymbolsSource = "sf-symbols"
  public static let windowsSource = "windows"
  public static let bootstrapSource = "bootstrap"
  public static let emojiSource = "emoji"

  public var sources: [String: [String: String]]

  public init(sources: [String: [String: String]] = [:]) {
    self.sources = sources
  }

  public init(tomlData: Data) throws {
    guard let text = String(data: tomlData, encoding: .utf8) else {
      throw BundleIconMapError.invalidLine(1, "Icon map file is not UTF-8.")
    }
    self.sources = try Self.parse(text)
  }

  public func value(for key: String?, source: String) -> String? {
    guard let key = key?.trimmingCharacters(in: .whitespacesAndNewlines), !key.isEmpty else {
      return nil
    }
    return sources[source]?[key]
  }

  public func resolving(_ key: String?, source: String, fallbackToKey: Bool = false) -> String? {
    value(for: key, source: source) ?? (fallbackToKey ? key?.nonEmpty : nil)
  }

  public func merging(_ overrides: BundleIconMap) -> BundleIconMap {
    var merged = sources
    for (source, values) in overrides.sources {
      merged[source, default: [:]].merge(values) { _, override in override }
    }
    return BundleIconMap(sources: merged)
  }

  private static func parse(_ text: String) throws -> [String: [String: String]] {
    var sources: [String: [String: String]] = [:]
    var currentSource: String?
    let lines = text.components(separatedBy: .newlines)

    for (offset, rawLine) in lines.enumerated() {
      let lineNumber = offset + 1
      let line = rawLine.trimmingCharacters(in: .whitespacesAndNewlines)
      if line.isEmpty || line.hasPrefix("#") {
        continue
      }

      if line.hasPrefix("[") && line.hasSuffix("]") {
        let source = String(line.dropFirst().dropLast())
          .trimmingCharacters(in: .whitespacesAndNewlines)
        guard !source.isEmpty else {
          throw BundleIconMapError.invalidLine(lineNumber, rawLine)
        }
        currentSource = source
        sources[source, default: [:]] = sources[source, default: [:]]
        continue
      }

      guard let source = currentSource,
        let equals = line.firstIndex(of: "=")
      else {
        throw BundleIconMapError.invalidLine(lineNumber, rawLine)
      }

      let rawKey = line[..<equals].trimmingCharacters(in: .whitespaces)
      let key = unquoteKey(String(rawKey))
      let rawValue = line[line.index(after: equals)...].trimmingCharacters(in: .whitespaces)
      sources[source, default: [:]][key] = try parseStringValue(
        rawValue,
        lineNumber: lineNumber,
        rawLine: rawLine)
    }

    return sources
  }

  private static func unquoteKey(_ key: String) -> String {
    if key.hasPrefix("\""), key.hasSuffix("\"") {
      return String(key.dropFirst().dropLast())
    }
    return key
  }

  private static func parseStringValue(
    _ rawValue: String,
    lineNumber: Int,
    rawLine: String
  ) throws -> String {
    guard rawValue.hasPrefix("\"") else {
      throw BundleIconMapError.invalidLine(lineNumber, rawLine)
    }
    let characters = Array(rawValue)
    var cursor = 1
    var closing: Int?
    while cursor < characters.count {
      let character = characters[cursor]
      if character == "\\" {
        cursor += 2
        continue
      }
      if character == "\"" {
        closing = cursor
        break
      }
      cursor += 1
    }
    guard let closing else {
      throw BundleIconMapError.invalidLine(lineNumber, rawLine)
    }
    let trailing = String(characters[(closing + 1)...]).trimmingCharacters(in: .whitespaces)
    if !trailing.isEmpty && !trailing.hasPrefix("#") {
      throw BundleIconMapError.invalidLine(lineNumber, rawLine)
    }
    return try unescape(String(characters[1..<closing]), lineNumber: lineNumber, rawLine: rawLine)
  }

  private static func unescape(
    _ value: String,
    lineNumber: Int,
    rawLine: String
  ) throws -> String {
    let characters = Array(value)
    var result = ""
    var index = 0
    while index < characters.count {
      let character = characters[index]
      guard character == "\\" else {
        result.append(character)
        index += 1
        continue
      }
      index += 1
      guard index < characters.count else {
        throw BundleIconMapError.invalidLine(lineNumber, rawLine)
      }
      let escaped = characters[index]
      index += 1
      switch escaped {
      case "n": result.append("\n")
      case "r": result.append("\r")
      case "t": result.append("\t")
      case "\"": result.append("\"")
      case "\\": result.append("\\")
      case "u", "U":
        let count = escaped == "u" ? 4 : 8
        guard index + count <= characters.count else {
          throw BundleIconMapError.invalidLine(lineNumber, rawLine)
        }
        let hex = String(characters[index..<(index + count)])
        guard let scalarValue = UInt32(hex, radix: 16),
          let scalar = UnicodeScalar(scalarValue)
        else {
          throw BundleIconMapError.invalidLine(lineNumber, rawLine)
        }
        result.unicodeScalars.append(scalar)
        index += count
      default:
        throw BundleIconMapError.invalidLine(lineNumber, rawLine)
      }
    }
    return result
  }
}

public enum BundleIconMapError: LocalizedError, Equatable {
  case invalidLine(Int, String)

  public var errorDescription: String? {
    switch self {
    case .invalidLine(let line, let text):
      "Invalid icon map TOML at line \(line): \(text)"
    }
  }
}
